bpfassets/libbpf/src/kepler.bpf.c

```diff
@@ -213,41 +270,38 @@ int kepler_sched_switch_trace(struct sched_switch_info *ctx)
 		counter_sched_switch = SAMPLE_RATE;
 	}
 
-	prev_pid = ctx->prev_pid;
-	pid_tgid = bpf_get_current_pid_tgid();
-	cur_pid = pid_tgid & 0xffffffff;
-	tgid = pid_tgid >> 32;
-	cgroup_id = bpf_get_current_cgroup_id();
-	cpu_id = bpf_get_smp_processor_id();
-	cur_ts = bpf_ktime_get_ns();
-	buf.cpu_cycles = get_on_cpu_cycles(&cpu_id);
-	buf.cpu_instr = get_on_cpu_instr(&cpu_id);
-	buf.cache_miss = get_on_cpu_cache_miss(&cpu_id);
-	buf.process_run_time = get_on_cpu_time(cur_pid, prev_pid, cur_ts);
-	buf.task_clock_time = buf.cpu_cycles / 1000000; // convert to ms
-
-	prev_pid_metrics = bpf_map_lookup_elem(&processes, &prev_pid);
-	if (prev_pid_metrics) {
-		// update process time
-		prev_pid_metrics->process_run_time += buf.process_run_time;
-		prev_pid_metrics->task_clock_time += buf.task_clock_time;
-		prev_pid_metrics->cpu_cycles += buf.cpu_cycles;
-		prev_pid_metrics->cpu_instr += buf.cpu_instr;
-		prev_pid_metrics->cache_miss += buf.cache_miss;
+	if (prev_state == TASK_RUNNING) {
+		// Skip if the previous thread was not registered yet
+		prev_tgid = bpf_map_lookup_elem(&pid_tgid_map, &prev_pid);
+		if (prev_tgid) {
+			// The process_run_time is 0 if we do not have the previous timestamp of
+			// the task or due to a clock issue. In either case, we skip collecting
+			// all metrics to avoid discrepancies between the hardware counter and CPU
+			// time.
+			if (buf.process_run_time > 0) {
+				prev_tgid_metrics = bpf_map_lookup_elem(
+					&processes, prev_tgid);
+				if (prev_tgid_metrics) {
```

Comment:
if we are aggregating power consumption for the whole process in bpf code only, do we need [`pid` and `tgid` separately](https://github.com/sustainable-computing-io/kepler/blob/54c6e85cb6af9e4e1d2b46053545f5cc9257fda4/bpfassets/libbpf/src/kepler.bpf.h#L94-L95) in `process_metrics_t`  ?

Reply:
was this aggregation also happening in user space kepler ?

Reply:
Good point, I have actually already mentioned that in another thread.
So, we don't need the `pid` and `tgid` in the `process_metrics_t`.
We can just have the `pid`, which will be the `tgid` from the kernel space.